litt/littdb_config.go

```diff
@@ -0,0 +1,196 @@
+	MaxSegmentKeyCount uint64
+
+	// The desired maximum size for a key file. The default is 2 MB. When a key file exceeds this size, the segment
+	// will close the current segment and begin writing to a new one. For workloads with moderately large values,
+	// this threshold is unlikely to be relevant. For workloads with very small values, this constant prevents a key
+	// file from growing too large. A key file with too many keys may have undesirable properties such as very slow
+	// garbage collection (since no kv-pair in a segment can be deleted until the entire segment is deleted).
+	TargetSegmentKeyFileSize uint64
+
+	// The period between garbage collection runs. The default is 5 minutes.
+	GCPeriod time.Duration
+
+	// The size of the keymap deletion batch for garbage collection. The default is 10,000.
+	GCBatchSize uint64
+
+	// The sharding factor for the database. The default is 8. Must be at least 1.
+	ShardingFactor uint32
+
+	// The random number generator used for generating sharding salts. The default is a standard rand.New()
+	// seeded by the current time.
+	SaltShaker *rand.Rand
+
+	// The size of the cache for tables that have not had their cache size set. The default is 0 (no cache).
+	// Cache size is in bytes, and includes the size of both the key and the value. Cache size can be set
+	// individually on each table by calling Table.SetCacheSize().
+	CacheSize uint64
+
+	// The time source used by the database. This can be substituted for an artificial time source
+	// for testing purposes. The default is time.Now.
+	TimeSource func() time.Time
+
+	// If true, then flush operations will call fsync on the underlying file to ensure data is flushed out of the
+	// operating system's buffer and onto disk. Setting this to false means that even after flushing data,
+	// there may be data loss in the advent of an OS/hardware crash.
+	//
+	// The default is true.
+	//
+	// Enabling fsync may have performance implications, although this strongly depends on the workload. For large
+	// batches that are flushed infrequently, benchmark data suggests that the impact is minimal. For small batches
+	// that are flushed frequently, the difference can be severe. For example, when enabled in unit tests that do
+	// super tiny and frequent flushes, the difference in performance was an order of magnitude.
+	Fsync bool
+
+	// If enabled, the database will return an error if a key is written but that key is already present in
+	// the database. Updating existing keys is illegal and may result in unexpected behavior, and so this check
+	// acts as a safety mechanism against this sort of illegal operation. Unfortunately, if using a keymap other
+	// than keymap.MemKeymapType, performing this check may be very expensive. By default, this is false.
+	DoubleWriteProtection bool
+
+	// If enabled, collect DB metrics and export them to prometheus. By default, this is false.
+	MetricsEnabled bool
```

Comment:
Is levelDB's approach less clean than embedding metric (levelDB makes stats available via API, so the metric collection runs outside)?

Reply:
The way levelDB does metrics is definitely more general purpose. If we ever decide to publish littDB as a stand alone utility intended for public use, I think this would be the proper way to handle metrics. The reason why it is currently tightly coupled with prometheus is primarily a simplicity thing. We currently don't have any intentions of supporting different metrics frameworks, and adding that capability would increase the scope of this PR. Willing to implement a wrapper like this now if you think it's important, but would prefer to do this as a future update.